### ISSUE TITLE
How to use ern cauldron add, when git repo is protected

### ISSUE THREAD

**Issue by jomnius** [electrode-io/electrode-native#615]
Not really a bug, but call for help:

In our environment it is NOT ALLOWED to write directly into git repo master branch. Everything must go through custom branch, reviewed and approved by at least two persons - for legal reasons. There is a watchdog to protect master branch from direct writes.

When I want to add new dependency into miniapp, I do it using "ern cauldron add dependencies" command, which (quite surprisingly) wants to push into cauldron git repo master branch. However local files are not changed. This is extremely confusing.

BUG 1:

Is there anything I can do to a) stop automatic hidden push into cauldron repo master branch? Or can I b) forward that into some other branch (which would be reviewed, approved and merged manually)?

BUG 2:

Now that I understand what is going on, I need to pull changes in order to see what actually happened. I guess there's nothing that can be automated here, just checking.

FIX:

Maybe just new documentation? I couldn't find any good (obvious) spot, where to write a PR suggestion...

**Comment by belemaire** [comment]
Hi @jomnius,

This is a bit of a bummer indeed :(

The Cauldron should be considered as a database. The reason we decided to use git instead of a fully fledged database is only for simplifying things and make it much more easy to setup for users (dedicating a machine to setup and host a fully fledged DB would have been a barrier of entry to Electrode Native for most users v.s the simplicity and speed of creating a git repo).

The problem we knew we would face by choosing git as the underlying DB store for Electrode Native would be that most users would see it as what it is : a git repo, and apply the same practices they are used to with git repos. Create a branch, issue a PR, review, merge. This should only be done for modifications that are manually made to the Cauldron (which should only happen for updating/adding configuration as we have no command for this yet). Otherwise if multiple Electrode Native clients are accessing the Cauldron they might rely on data which is not up date (waiting to be merged ?) and it will create big desynchronization issues.

Anyway on your end this is more problematic as it is a technical restriction. Even without going through PR/review/merge cycle (which you shouldn't go through for Cauldron unless it is a manual change), you are blocked because of the master branch protection.

If Electrode Native client was to work of another branch, it would be able to properly push to this branch. The only thing is how to properly ensure all clients of your Cauldron use the correct branch.

Here is what we could do, please let me know if you are OK with this solution, if so we will create an appropriate issue to implement this :

- Modify `ern cauldron repo add` command logic so that it allows to provide a GitHub repository url including a branch (following standard format). For example `ern cauldron repo add my-cauldron https://github.com/belemaire/dev-cauldron#mybranch`. If the branch is omitted it will assume `master` by default, otherwise it will work of the specified branch (`mybranch` in this case). This way it makes it easy for every user of Electrode Native within an org to specify the proper branch when adding the Cauldron repo to their local Electrode Native client. Any command that is accessing the Cauldron will use the specific branch if any, for the Cauldron being accessed.

We will also need to do something similar for GitHub publication of Containers (in the case you are using a GitHub publisher to publish your Container(s)), as it also push the Containers to the master branch. In a similar way to Cauldron, if a branch is specified in the publisher configuration in the Cauldron, for example :

```json
"publishers": [
 {
     "name": "github",
     "url": "https://github.com/belemaire/my-container.git#mybranch"
}
``` 

it will use the specified branch instead of pushing to master by default.

I think this will solve the problem you are facing. Please let me know if that would be ok for you, or if you have other suggestions concerns. This solution shouldn't be too difficult to implement and we could have it released with next release of the platform on Monday

**Comment by jomnius** [comment]
Thanx for quick reply, need to think about it! However one comment immediately: as a lazy developer I wouldn't feel comfortable with typing long branch names. I'll get it wrong sooner than later, therefore I would have to automate it :)

SUGGESTION:

Could that automation be built-in? Since we're using git-flow, I have a local branch and be working in it. Therefore maybe the default branch name could be "current" branch, if nothing else is defined? For example:

```bash
jmiettunen@mac my-miniapp $ git rev-parse --abbrev-ref HEAD
feature/fix_without_build
```

The start of the url you already got from somewhere, I haven't checked? In any case something like "git remote get-url origin" might help? Btw using GitHub Enterprise with ssh auth, if it makes any difference:

```bash
jmiettunen@mac my-miniapp $ git remote get-url origin
git@github.bus.compa.ny:Project/my-miniapp.git
```

COMMENTS:

I might be able to define cauldron repo as "docs" with your definition of how system works. The problem is still occasional manual commits, which is the case where direct "source code" master branch commits are not allowed for legal reasons.

Being out-of-sync is not something I would worry about. Master branch should always work as-is. Of course sometimes our "default" branch has been develop, btw. If that makes any difference.

Actually we already have problems with "force push to master", since we're working on 2 separate mini apps and while someone is working on feature branch, they mess up others' setup by modifying master manifest/cauldron, while the rest of the code is still not merged. Just to let you know.

GitHub publication of Containers

Not that far yet, but I believe you're right. My main interest would be anyway to do a really clean build from all original source codes at the same time.

Currently we have only "maven" publisher for android. Releasing and using iOS binaries is not something I'm interested in. I have used Carthage previously to integrate iOS binary frameworks from GitHub and it was nothing but problems. Might be useful for others, happy for them!

**Comment by belemaire** [comment]
@jomnius 
I have created issue https://github.com/electrode-io/electrode-native/issues/631 so that we can work on it for next Monday release. Even if that might not be the ideal final solution (if such one exist), at least it will unblock you and potential future users that might face the same issue.

The Cauldron is accessed by Electrode Native for some commands as read-only, but as you have noticed, some commands are also pushing to the Cauldron. Whenever you issue a command that is going to modify the internal state of a Container (for example adding/removing/updating dependencies, miniapps), then a new Container needs to be generated (and published) to contain the changes. The version of the current Container is also properly updated in the Cauldron. The Cauldron is therefore guaranteed to reflect the current state of a native application version Container (the latest version of the Container and every dependencies and miniapps it contains). 

Once again it's a DB and desynchronization can have bad consequences. For example if you work from a fork of the 'master' cauldron and add a MiniApp to a native application version (with current version of the Container set to 1.0.0 in the Cauldron let's say) , it will regenerate a Container and update your Cauldron fork with the new Container version (1.0.1). If you take some time to merge your branch to 'master' cauldron, and meanwhile some other team member is now doing the same (adding a MiniApp or updating a dependency or what else) from it's own fork, then Electrode Native doesn't have information that latest Container version is `1.0.1` and not `1.0.0`. It will regenerate a Container with version `1.0.1` and when it tries to publish it, it will fail as version `1.0.1` was already published to Maven (or GitHub if you are using GitHub publication). That's just one of the reason why it's not a good idea to work from forks of Cauldrons (once again just needed for manual changes to the Cauldron, only Config). The Cauldron is the database used by Electrode Native and is assumed to be consistent and up to date at any time.

We will have more information in the documentation soon, regarding CI as we have almost finalized our CI workflow, this might be of help to you as well.

**Comment by belemaire** [comment]
Released in v0.14.0.
Closing this issue now.
